UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c

```diff
@@ -231,13 +257,236 @@ SmmMemoryProtectionsTestInvalidRange (
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting 32-bit I/O read from non-authorized port 0xCF8...\n", __FUNCTION__));
+
+  ReadData = IoRead32(0xCF8);
+
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Unauthorized I/O write test.
+  This test will attempt to write a non-authorized I/O port from SMM.
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestUnathorizedIoWrite (
+  VOID
+  )
+{
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting 32-bit I/O write to non-authorized port 0xCF8...\n", __FUNCTION__));
+
+  IoWrite32(0xCF8, 0x80000000);
+
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Unauthorized MSR read test.
+  This test will attempt to read a non-authorized MSR from SMM.
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestUnathorizedMsrRead (
+  VOID
+  )
+{
+  UINT64 ReadData;
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting to read non-authorized MSR IA32_MTRR_PHYSBASE0...\n", __FUNCTION__));
+
+  ReadData = AsmReadMsr64(MSR_IA32_MTRR_PHYSBASE0);
+
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Unauthorized MSR write test.
+  This test will attempt to write a non-authorized MSR from SMM.
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestUnathorizedMsrWrite (
+  VOID
+  )
+{
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting to write non-authorized MSR IA32_MTRR_PHYSBASE0...\n", __FUNCTION__));
+
+  AsmWriteMsr64(MSR_IA32_MTRR_PHYSBASE0, 0);
```

Comment:
i wonder if all the values of "unauthorized*" should be external in header file or maybe even library.  As i can image some differences between silicon partners.